samza-api/src/main/java/org/apache/samza/system/StreamSpec.java

```diff
@@ -0,0 +1,197 @@
+   */
+  public StreamSpec(String id, String physicalName, String systemName, int partitionCount) {
+    this(id, physicalName, systemName, partitionCount, Collections.emptyMap());
+  }
+
+  /**
+   *  @param id           The application-unique logical identifier for the stream. It is used to distinguish between
+   *                      streams in a Samza application so it must be unique in the context of one deployable unit.
+   *                      It does not need to be globally unique or unique with respect to a host.
+   *
+   * @param physicalName  The physical identifier for the stream. This is the identifier that will be used in remote
+   *                      systems to identify the stream. In Kafka this would be the topic name whereas in HDFS it
+   *                      might be a file URN.
+   *
+   * @param systemName    The System name on which this stream will exist. Corresponds to a named implementation of the
+   *                      Samza System abstraction. See {@link SystemFactory}
+   *
+   * @param config        A map of properties for the stream. These may be System-specfic.
+   */
+  public StreamSpec(String id, String physicalName, String systemName, Map<String, String> config) {
+    this(id, physicalName, systemName, DEFAULT_PARTITION_COUNT, config);
+  }
+
+  /**
+   *  @param id             The application-unique logical identifier for the stream. It is used to distinguish between
+   *                        streams in a Samza application so it must be unique in the context of one deployable unit.
+   *                        It does not need to be globally unique or unique with respect to a host.
+   *
+   * @param physicalName    The physical identifier for the stream. This is the identifier that will be used in remote
+   *                        systems to identify the stream. In Kafka this would be the topic name whereas in HDFS it
+   *                        might be a file URN.
+   *
+   * @param systemName      The System name on which this stream will exist. Corresponds to a named implementation of the
+   *                        Samza System abstraction. See {@link SystemFactory}
+   *
+   * @param partitionCount  The number of partitionts for the stream. A value of {@code 1} indicates unpartitioned.
+   *
+   * @param config          A map of properties for the stream. These may be System-specfic.
+   */
+  public StreamSpec(String id, String physicalName, String systemName, int partitionCount,  Map<String, String> config) {
+    if (id == null) {
+      throw new NullPointerException("Parameter 'id' must not be null");
+    }
+
+    if (systemName == null) {
+      throw new NullPointerException("Parameter 'systemName' must not be null");
+    }
+
+    this.id = id;
+    this.systemName = systemName;
+    this.physicalName = physicalName;
+    this.partitionCount = partitionCount;
+
+    if (config != null) {
+      this.config = Collections.unmodifiableMap(config);
```

Comment:
You also need to copy the map here. Otherwise it would be possible to modify the map if you had a reference to the backing map.

Something like this would work:

```
this.config = Collections.unmodifiableMap(new HashMap<String, String>(config));
```